Add commit_time metric in pg_stat_database

lpwisniewski opened the pull request.

bonnefoa commented on src/backend/catalog/system_views.sql:
```diff
@@ -1084,7 +1084,8 @@ CREATE VIEW pg_stat_database AS
             pg_stat_get_db_parallel_workers_to_launch(D.oid) as parallel_workers_to_launch,
             pg_stat_get_db_parallel_workers_launched(D.oid) as parallel_workers_launched,
-            pg_stat_get_db_stat_reset_time(D.oid) AS stats_reset
+            pg_stat_get_db_stat_reset_time(D.oid) AS stats_reset,
+            pg_stat_get_db_commit_time(D.oid) AS commit_time
```
stats_reset should stay as the last column by (implicit?) convention

bonnefoa commented on src/backend/utils/activity/pgstat_xact.c:
```diff
@@ -37,9 +38,17 @@ static PgStat_SubXactStatus *pgStatXactStack = NULL;
+AtEOXact_PgStat(bool isCommit, bool parallel, instr_time start_time)
 {
 	PgStat_SubXactStatus *xact_state;
+	instr_time now;
+	instr_time elapsed;
+
+	INSTR_TIME_SET_ZERO(elapsed);
+	INSTR_TIME_SET_CURRENT(now);
+	INSTR_TIME_ACCUM_DIFF(elapsed, now, start_time);
```
This could be simplified to
```
instr_time duration;

INSTR_TIME_SET_CURRENT(duration);
INSTR_TIME_SUBSTRACT(duration, start_time);
```

bonnefoa commented on src/backend/utils/adt/pgstatfuncs.c:
```diff
@@ -1204,6 +1204,9 @@ PG_STAT_GET_DBENTRY_FLOAT8_MS(blk_write_time)
 /* pg_stat_get_db_idle_in_transaction_time */
 PG_STAT_GET_DBENTRY_FLOAT8_MS(idle_in_transaction_time)
 
+/* pg_stat_get_db_idle_in_transaction_time */
```
Should be changed to pg_stat_get_db_commit_time

bonnefoa commented on src/include/catalog/pg_proc.dat:
```diff
@@ -12433,6 +12433,11 @@
   proargtypes => 'regclass name bool',
   proargnames => '{relation,attname,inherited}',
   prosrc => 'pg_clear_attribute_stats' },
+{ oid => '9543',
+  descr => 'statistics: commit time, in milliseconds',
+  proname => 'pg_stat_get_db_commit_time', provolatile => 's',
+  proparallel => 'r', prorettype => 'float8', proargtypes => 'oid',
+  prosrc => 'pg_stat_get_db_commit_time' },
```
You should keep this grouped with the other pg_stat_get functions, just after the last pg_stat_get_db_parallel_workers_launched

bonnefoa commented on src/backend/access/transam/twophase.c:
```diff
@@ -197,6 +197,12 @@ static GlobalTransaction MyLockedGxact = NULL;
+ * xactCommitStartTime is the time at which we started to commit
+ * the current transaction.
+ */
+static instr_time xactCommitStartTime;
```
Maybe a better place for this would be in pgstat_xact. You keep a `pgStatPreCommitTime` as a global variable and expose a `AtPreCommit_PgStat` function that would set this variable to the current time.
This way, you will just need to call `AtPreCommit_PgStat` in CommitTransaction (and the equivalent for the 2PC) and `AtEOXact_PgStat` can directly use `pgStatPreCommitTime`

bonnefoa reviewed (commented):
Catalog version in src/include/catalog/catversion.h needs to be increased.
The new metric should also be documented in doc/src/sgml/monitoring.sgml

lpwisniewski committed:
Add commit_time metric in pg_stat_database

Creates a new metric in pg_stat_database to measure the transaction commit duration.
It changes xact.c and twophase.c to track the commit start time then relies on AtEOXact_PgStat
to pass the calculated duration to pg_stat_database variables.

bonnefoa commented on src/backend/access/transam/twophase.c:
```diff
@@ -1503,6 +1505,8 @@ FinishPreparedTransaction(const char *gid, bool isCommit)
 	xl_xact_stats_item *abortstats;
 	SharedInvalidationMessage *invalmsgs;
 
+	pgstat_report_commit_start();
```
This is only needed if isCommit is true?

bonnefoa commented on src/backend/access/transam/xact.c:
```diff
@@ -70,6 +70,7 @@
 #include "utils/timeout.h"
 #include "utils/timestamp.h"
 #include "utils/typcache.h"
+#include "portability/instr_time.h"
```
This include is not necessary anymore

bonnefoa commented on src/backend/utils/activity/pgstat_xact.c:
```diff
@@ -16,6 +16,7 @@
 #include "pgstat.h"
 #include "utils/memutils.h"
 #include "utils/pgstat_internal.h"
+#include "portability/instr_time.h"
```
No need for this include

bonnefoa commented on src/backend/utils/activity/pgstat_database.c:
```diff
@@ -201,6 +204,15 @@ pgstat_report_connect(Oid dboid)
+ * Register the last commit start time.
+ */
+void
+pgstat_report_commit_start(void)
```
The name isn't great. pgstat_report* functions usually modify the shared memory. Here, we only track the start.
Maybe a better name would be PreCommit_PgStat to do the pre commit part of AtEOXact_PgStat and could accept a isCommit which can be used by the 2pc part.

lpwisniewski replied:
I tried to find some naming convention for function to know when should a function is capitalised but I wasn't able to find anything. Is there any strict rule around it?

lpwisniewski replied:
I've changed it let me know what you think

bonnefoa commented on src/backend/utils/activity/pgstat_database.c:
```diff
@@ -29,12 +30,14 @@ PgStat_Counter pgStatBlockReadTime = 0;
 static int	pgStatXactCommit = 0;
 static int	pgStatXactRollback = 0;
 static PgStat_Counter pgLastSessionReportTime = 0;
+static PgStat_Counter pgLastStartCommitTime = 0;
```
This should be a TimestampTz instead (yeah, pgLastSessionReportTime should also be a TimestampTz, but that should be done in a different patch.

bonnefoa commented on src/include/pgstat.h:
```diff
@@ -622,6 +624,8 @@ extern void pgstat_update_parallel_workers_stats(PgStat_Counter workers_to_launc
 	(pgStatActiveTime += (n))
 #define pgstat_count_conn_txn_idle_time(n)							\
 	(pgStatTransactionIdleTime += (n))
+#define pgstat_commit_time(n)							\
```
Name should be `pgstat_count_commit_time`, similar to what's done with `pgstat_count_conn_txn_idle_time`

bonnefoa commented on src/backend/utils/activity/pgstat_database.c:
```diff
@@ -256,7 +272,13 @@ AtEOXact_PgStat_Database(bool isCommit, bool parallel)
 			pgStatXactCommit++;
+			now = GetCurrentTimestamp();
+
+			TimestampDifference(pgLastStartCommitTime, now, &secs, &usecs);
```
You can directly use TimestampDifferenceMilliseconds, this would save you the need to do the sec+usecs conversion

bonnefoa commented on src/backend/utils/activity/pgstat_database.c:
```diff
@@ -20,6 +20,7 @@
 #include "storage/procsignal.h"
 #include "utils/pgstat_internal.h"
 #include "utils/timestamp.h"
+#include "pgstat.h"
```
This is not needed I think.

bonnefoa commented:
Also, you need to add a newline between the commit title (first line) and the message body.
